setup.py

```diff
@@ -28,30 +28,26 @@ def get_version():
     url='https://github.com/Shopify/pyoozie',
     packages=['pyoozie'],
     install_requires=[
-        'enum34>=0.9.23',
-        'requests>=2.12.3',
-        'six>=1.10.0',
-        'typing',
-        'untangle>=1.1.0',
-        'yattag>=1.7.2',
+        'enum34 >= 0.9.23',
+        'requests >= 2.12.3',
+        'six >= 1.10.0',
+        'typing >= 3.6.1',
+        'untangle >= 1.1.0',
+        'yattag >= 1.7.2',
     ],
     extras_require={
-        'deploy': [
-            'setuptools>=0.9'
-        ],
         'test': [
-            'autopep8',
-            'flake8',
-            'mock',
-            'mypy; python_version >= "3.3"',
+            'autopep8 == 1.3.1',
+            'mock == 2.0.0; python_version < "3.3"',
+            'mypy == 0.501; python_version >= "3.3"',
             'pycodestyle == 2.2.0',
-            'pylint',
-            'pytest-cov',
-            'pytest-randomly',
-            'pytest>=2.7',
-            'requests-mock',
-            'shopify_python==0.2.2',
-            'xmltodict',
+            'pylint == 1.6.5',
```

Comment:
Sorry, thought about it some more and I'm not sure this makes sense. You are explicitly setting the versions of your immediate dependencies, but will you also list all of your transitive dependencies? If not, you are still susceptible to non-deterministic builds.

If you want really deterministic builds I think you need to actually install the packages, let Pip pick whatever it does, and then freeze it to a requirements file. I would argue that it's not a requirement. In fact, you might as well have something that will fail (and tell you) if/when a new version of a dependency is released (that meets your production requirements list) that doesn't pass your tests.